Fix request->is(xml) returning true for HTML requests

markstory opened the pull request:
Add html/text as an option for all content-type based detectors so that we can differentiate between requests that want HTML but also take xml or json from being confused as a non-html type. Assuming browser based traffic is a safe default in my opinion.

Fixes #16583

markstory committed:
Fix request->is(xml) returning true for HTML requests

Add html/text as an option for all content-type based detectors so that
we can differentiate between requests that want HTML but also take xml
or json from being confused as a non-html type. Assuming browser based
traffic is a safe default in my opinion.

Fixes #16583

othercorey commented on src/Http/ServerRequest.php:
```diff
@@ -553,9 +553,12 @@ protected function _is(string $type, array $args): bool
         $content = new ContentTypeNegotiation();
-        $accepted = $content->preferredType($this, $detect['accept']);
+        $options = $detect['accept'];
+        // We add text/html as the browsers often combine xml + html types together.
```
I think this needs a clearer comment about the effect we want to have.

markstory replied:
In writing a better explanation I realized that this change would break detecting HTML types. I've change the implementation around a bit and used a more explicit ignore/exclude list instead.

othercorey replied:
So we're saying if you have a request with html and xml accept headers and have only an xml renderer, then this isn't a valid xml request.

markstory replied:
> So we're saying if you have a request with html and xml accept headers and have only an xml renderer, then this isn't a valid xml request.

Kind of. It depends on the header value. An Accept header of  `text/html,application/xml` will not count as XML. As both XML and HTML are of equal preference value.

However, a request of `text/html;q=0.8, application/xml` *will* count as XML as HTML is a lower preference value. Given that many browsers include `text/html,application/xml;q=0.9` in their headers we can't only check if XML is accepted because while the browser will accept XML, it prefers HTML more.

Adding in a special case for the XML type seemed like the safest change to me. The previous implementation of `_acceptDetector` was:

```php
    protected function _acceptHeaderDetector(array $detect): bool
    {
        $acceptHeaders = explode(',', (string)$this->getEnv('HTTP_ACCEPT'));
        foreach ($detect['accept'] as $header) {
            if (in_array($header, $acceptHeaders, true)) {
                return true;
            }
        }

        return false;
    }
```

This logic was doing a strict check on only the unqualified types and `application/xml` != `application/xml;q=0.9`.

othercorey replied:
Ok, that makes sense to me. We're marking exclusions at equal preference. Instead of writing a sorting order, we're special casing exclusions. Other requests with matching equality get what they get from the server.

markstory committed:
Make content type detection more robust.

Someone could have a detector that wants to find text/html and the
previous change would break that.

othercorey commented on tests/TestCase/Http/ContentTypeNegotiationTest.php:
```diff
@@ -26,6 +26,21 @@ public function testPreferredTypeNoAccept()
+            ],
+        ]);
+        $this->assertEquals('text/html', $content->preferredType($request));
+        $this->assertEquals('text/html', $content->preferredType($request, ['text/html', 'application/json']));
```
Shouldn't this be testing with html and xml to ensure html wins with xml at 0.9?

I don't know that json is helpful here.

markstory committed:
Fix phpcs

ADmad commented on src/Http/ServerRequest.php:
```diff
@@ -553,9 +558,26 @@ protected function _is(string $type, array $args): bool
+        // Some detectors overlap with the default browser Accept header
+        // For these types we use an exclude list to refine our content type
+        // detection.
+        $exclude = null;
+        if (!empty($detect['exclude'])) {
+            $exclude = $detect['exclude'];
```
```suggestion
        $exclude = $detect['exclude'] ?: null;
        if ($exclude) {
```

markstory committed:
Apply suggestions from code review

Co-authored-by: ADmad <ADmad@users.noreply.github.com>

markstory committed:
Use ?? instead of ?:
